### ISSUE TITLE
NaViT is slower than expected

### ISSUE THREAD

**Issue by AmitMY** [lucidrains/vit-pytorch#347]
Full reproduction: https://github.com/sign/image-latent-transformer/blob/main/examples/benchmark_image_encoder.py

I am trying to find a model to encode lots of small images per batch.

I started by using Huggingface Transformers, and creating a `ViTModel`:
```py
ViTModel.from_pretrained("WinKawaks/vit-tiny-patch16-224")
```

However, this model does not know how to work on variable size images, and if padded, returns different results.

So I created a patch, that allows the model to accept padded images - https://github.com/sign/image-latent-transformer/blob/main/image_latent_transformer/vision/batch_image_encoder.py - and that works, but is ugly and can only support eager attention.

I moved to `vit-pytorch` and wrapped a `NaViT` model from this library with the huggingface setup:

<details>

<summary>NaViT Model</summary>

```py
import torch
from transformers import AutoModel, AutoConfig
from transformers.configuration_utils import PretrainedConfig
from transformers.modeling_outputs import BaseModelOutputWithPooling
from transformers.modeling_utils import PreTrainedModel
from vit_pytorch.na_vit_nested_tensor import NaViT


class NaViTConfig(PretrainedConfig):
    model_type = "navit"

    def __init__(
            self,
            image_size: int = 256,  # only used to set a default; NaViT handles var-size
            patch_size: int = 16,
            hidden_size: int = 512,
            dim: int = 512,
            depth: int = 6,
            heads: int = 8,
            mlp_dim: int = 1024,
            dropout: float = 0.0,
            emb_dropout: float = 0.0,
            token_dropout_prob: float = 0.1,
            **kwargs,
    ):
        super().__init__(**kwargs)
        self.image_size = image_size
        self.patch_size = patch_size
        self.hidden_size = hidden_size
        self.dim = dim
        self.depth = depth
        self.heads = heads
        self.mlp_dim = mlp_dim
        self.dropout = dropout
        self.emb_dropout = emb_dropout
        self.token_dropout_prob = token_dropout_prob


class NaViTModel(PreTrainedModel):
    config_class = NaViTConfig

    def __init__(self, config: NaViTConfig):
        super().__init__(config)

        self.navit = NaViT(
            image_size=config.image_size,
            patch_size=config.patch_size,
            num_classes=config.hidden_size,
            dim=config.dim,
            depth=config.depth,
            heads=config.heads,
            mlp_dim=config.mlp_dim,
            dropout=config.dropout,
            emb_dropout=config.emb_dropout,
            token_dropout_prob=config.token_dropout_prob,
        )

        # Initialize weights via HF utility (won't disturb vit-pytorch-initialized weights unless uninitialized)
        self.post_init()

    def forward(self, images: list[torch.Tensor], **kwargs):
        # vit-pytorch NaViT returns shape (B, num_classes)
        logits = self.navit(images)

        return BaseModelOutputWithPooling(
            last_hidden_state=None,
            pooler_output=logits
        )

AutoConfig.register(NaViTConfig.model_type, NaViTConfig)
AutoModel.register(NaViTConfig, NaViTModel)
```
</details>

I created three models of similar size, and then benchmarked the various setups:
```py
    vit_model = ViTModel.from_pretrained("WinKawaks/vit-tiny-patch16-224")

    patched_vit_model = ViTModel.from_pretrained("WinKawaks/vit-tiny-patch16-224")
    maybe_patch_vit_model(vit_model)

    # Similar model~
    navit_model = NaViTModel(NaViTConfig(
        image_size=512,  # Max size for pos embeddings
        patch_size=vit_model.config.patch_size,
        hidden_size=vit_model.config.hidden_size,
        dim=vit_model.config.intermediate_size,
        depth=vit_model.config.num_hidden_layers // 2,
        heads=vit_model.config.num_attention_heads,
        mlp_dim=128,
    ))
```

Batch size: 15k images (size on average 16x64~)
GPU: A100

1. ViT Model with padding (inconsistent results) - 1 second
2. Vit Model grouped into batches of same size images - 13 seconds
3. Vit Model patched (consistent results) - 34 second
4. NaVit Model - 39 seconds

**Comment by AmitMY** [comment]
Now that #353 is merged. Here's a better benchmark script with more images (512 variable-width images (16px tall, 32-80px wide))

```py
import time
import statistics
import random
import torch
from vit_pytorch.vit import ViT
from vit_pytorch.na_vit import NaViT

NUM_RUNS = 10  # Number of benchmark runs for statistics
ITERS_PER_RUN = 10  # Iterations per run

# 512 variable-width images: 16px tall (1 patch), 32-80px wide (2-5 patches)
random.seed(42)
widths = [random.choice([32, 48, 64, 80]) for _ in range(512)]
images = [torch.randn(3, 16, w) for w in widths]

device = "cuda" if torch.cuda.is_available() else "cpu"

# Models with same architecture (image_size=80 to handle largest width)
kwargs = dict(image_size=80, patch_size=16, num_classes=100, dim=192, depth=12, heads=3, mlp_dim=768)
vit = ViT(**kwargs).to(device).eval()
navit = NaViT(**kwargs, token_dropout_prob=None).to(device).eval()

# Prepare inputs
# For ViT: pad all images to max size (16, 80)
max_h, max_w = 16, 80
padded_images = []
for img in images:
    h, w = img.shape[1], img.shape[2]
    padded_img = torch.zeros(3, max_h, max_w)
    padded_img[:, :h, :w] = img
    padded_images.append(padded_img)
padded = torch.stack(padded_images).to(device)  # (512, 3, 16, 80)
images_gpu = [img.to(device) for img in images]


def benchmark(fn, warmup=5):
    """Run benchmark and return time per batch in ms."""
    # Warmup
    for _ in range(warmup):
        fn()
    if device == "cuda":
        torch.cuda.synchronize()

    # Timed run
    start = time.perf_counter()
    for _ in range(ITERS_PER_RUN):
        fn()
    if device == "cuda":
        torch.cuda.synchronize()
    elapsed = time.perf_counter() - start
    return elapsed * 1000 / ITERS_PER_RUN  # ms per batch


# Collect results across multiple runs
vit_times = []
navit_times = []

print(f"Running {NUM_RUNS} benchmark runs ({ITERS_PER_RUN} iters each)...")
print(f"Config: 512 variable-width images (16px tall, 32-80px wide)")

with torch.no_grad():
    for run in range(NUM_RUNS):
        vit_times.append(benchmark(lambda: vit(padded)))
        navit_times.append(benchmark(lambda: navit(images_gpu)))
        print(f"  Run {run + 1}/{NUM_RUNS} done")

# Compute statistics
def stats(times):
    return {
        "mean": statistics.mean(times),
        "std": statistics.stdev(times) if len(times) > 1 else 0,
        "min": min(times),
        "max": max(times),
    }

vit_stats = stats(vit_times)
navit_stats = stats(navit_times)

def fmt_time(mean, std):
    return f"{mean:.1f}±{std:.1f}ms"

def fmt_ratio(mean, std, baseline_mean):
    ratio = mean / baseline_mean
    ratio_std = std / baseline_mean
    if ratio < 1.05:
        return "1x"
    return f"{ratio:.1f}±{ratio_std:.1f}x slower"

vit_time = fmt_time(vit_stats['mean'], vit_stats['std'])
navit_time = fmt_time(navit_stats['mean'], navit_stats['std'])

navit_ratio = fmt_ratio(navit_stats['mean'], navit_stats['std'], vit_stats['mean'])

print()
print(f"| {'Model':<15} | {'Time per batch':^14} | {'vs ViT':^14} |")
print(f"|-{'-'*15}-|-{'-'*14}-|-{'-'*14}-|")
print(f"| {'ViT (padded)':<15} | {vit_time:^14} | {'1x (baseline)':^14} |")
print(f"| {'NaViT':<15} | {navit_time:^14} | {navit_ratio:^14} |")
```

```
pip install "vit-pytorch==1.16.2"
```
| Model           | Time per batch |     vs ViT     |
|-----------------|----------------|----------------|
| ViT (padded)    |   6.2±0.1ms    | 1x (baseline)  |
| NaViT           |   48.7±1.5ms   | 7.8±0.2x slower |
```
pip install "vit-pytorch==1.16.3"
```
| Model           | Time per batch |     vs ViT     |
|-----------------|----------------|----------------|
| ViT (padded)    |   6.2±0.0ms    | 1x (baseline)  |
| NaViT           |   32.6±0.4ms   | 5.3±0.1x slower |
